pkg/controller/remotemachineset/awsactuator.go

```diff
@@ -204,3 +253,73 @@ func (a *AWSActuator) updateProviderConfig(machineSet *machineapi.MachineSet, in
 		Value: &runtime.RawExtension{Object: providerConfig},
 	}
 }
+
+// getSubnetsByAvailabilityZones maps availability zones to subnet
+func (a *AWSActuator) getSubnetsByAvailabilityZone(pool *hivev1.MachinePool) (map[string]string, error) {
+	idPointers := make([]*string, len(pool.Spec.Platform.AWS.Subnets))
+	for i, id := range pool.Spec.Platform.AWS.Subnets {
+		idPointers[i] = aws.String(id)
+	}
+
+	results, err := a.awsClient.DescribeSubnets(&ec2.DescribeSubnetsInput{SubnetIds: idPointers})
+	if err != nil {
+		if strings.Contains(err.Error(), "InvalidSubnetID.NotFound") {
+			// formatting error message before adding it to condition
+			// sample error message: InvalidSubnetID.NotFound: The subnet ID 'subnet-1,subnet-2' does not exist status code: 400, request id: f506c630-ab03-4d81-b116-745e0b3e36ce
+			// message after formatting: The subnet ID 'subnet-1,subnet-2' does not exist
+			reg := regexp.MustCompile("\tstatus code:(.)*")
+			conditonMessage := strings.TrimPrefix(reg.ReplaceAllString(err.Error(), ""), "InvalidSubnetID.NotFound: ")
+
+			conds, changed := controllerutils.SetMachinePoolConditionWithChangeCheck(
+				pool.Status.Conditions,
+				hivev1.InvalidSubnetsMachinePoolCondition,
+				corev1.ConditionTrue,
+				"SubnetsNotFound",
+				conditonMessage,
+				controllerutils.UpdateConditionIfReasonOrMessageChange,
+			)
+			if changed {
+				pool.Status.Conditions = conds
+				if err := a.client.Status().Update(context.Background(), pool); err != nil {
+					return nil, err
+				}
+			}
+		}
+		return nil, err
+	}
+
+	conflictingSubnets := sets.NewString()
+	subnetsByAvailabilityZone := make(map[string]string, len(pool.Spec.Platform.AWS.Zones))
```

Comment:
Use the length of the subnets rather than the length of the availability zones. The availability zones slice may be empty. The happy scenario has the length of the subnets equal to the number of availability zones used.